test/operator-framework-e2e/operator_framework_test.go

```diff
@@ -0,0 +1,975 @@
+	}
+
+	// Check if bundle manifests are created
+	bundleManifestPath := filepath.Join(sdkInfo.projectName, "bundle", "manifests")
+	Expect(bundleManifestPath).To(BeAnExistingFile())
+
+	// Check if bundle metadata is created
+	bundleMetadataPath := filepath.Join(sdkInfo.projectName, "bundle", "metadata")
+	Expect(bundleMetadataPath).To(BeAnExistingFile())
+
+	// Build the bundle image
+	bundleImg := "BUNDLE_IMG=" + bundleData.imageRef
+	cmd = exec.Command("make", "bundle-build", bundleImg)
+	cmd.Dir = sdkInfo.projectName
+	output, err = cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("Error building bundle image %v with tag %v :%v: %v", bundleData.imageRef, bundleData.bundleVersion, string(output), err)
+	}
+
+	// Push the bundle image
+	cmd = exec.Command("make", "bundle-push", bundleImg)
+	cmd.Dir = sdkInfo.projectName
+	output, err = cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("Error pushing bundle image %v with tag %v : %v: %v", bundleData.imageRef, bundleData.bundleVersion, string(output), err)
+	}
+
+	// Load the bundle image into test environment
+	if err = loadImages(GinkgoWriter, kindServer, bundleData.imageRef); err != nil {
+		return err
+	}
+
+	// Move the bundle structure into correct testdata folder for bundles
+	err = moveFolderContents(filepath.Join(sdkInfo.projectName, "bundle"), filepath.Join(rootBundlePath, bundleData.bInputDir))
+	Expect(err).ToNot(HaveOccurred())
+	Expect(filepath.Join(rootBundlePath, bundleData.bInputDir)).To(BeAnExistingFile())
+
+	// Move the generated dockerfile to correct path
+	err = os.Rename(filepath.Join(sdkInfo.projectName, "bundle.Dockerfile"), filepath.Join(rootBundlePath, bundleData.bInputDir, "bundle.Dockerfile"))
+	Expect(err).ToNot(HaveOccurred())
+
+	return nil
+}
+
+// buildPushLoadContainer function builds a Docker container image using the Docker
+// command-line tool, pushes to a container registry and loads into a kind cluster.
+//
+// The function takes the following arguments and returns back the error if any:
+// `tag`: container image tag/name.
+// `dockerfilePath`: path to the Dockerfile that defines the container image.
+// `dockerContext`: context directory containing the files and resources referenced by the Dockerfile.
+// `w`: Writer to which the standard output and standard error will be redirected.
+func buildPushLoadContainer(tag, dockerfilePath, dockerContext, kindServer string, w io.Writer) error {
+	cmd := exec.Command("docker", "build", "-t", tag, "-f", dockerfilePath, dockerContext)
+	cmd.Stderr = w
+	cmd.Stdout = w
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("Error building Docker container image %s : %v", tag, err)
+	}
+
+	cmd = exec.Command("docker", "push", tag)
+	cmd.Stderr = w
+	cmd.Stdout = w
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("Error pushing Docker container image: %s to the registry: %v", tag, err)
```

Comment:
I'm gettign this error:
```
Error pushing Docker container image: localhost:5001/plain-catalog:test to the registry: exit status 125
```

I use `podman` with an alias from `docker` to `podman`.

I changed the local port from 5000 to 5001 due to issue with control center on macos mentioned above, but still fails.

Reply:
The test ran successfully on my macos. So I guess I am missing out something here. A local registry was used over the container registry `quay.io` for improved performance. What could be an ideal solution for the issue with the port?

Reply:
I'm on version Ventura 13.4.1 right now. It looks like it started listening on port 5000 starting from Monterey.

> What could be an ideal solution for the issue with the port?

I would change the localhost port to something else other than 5000. For example, [kind suggests](https://kind.sigs.k8s.io/docs/user/local-registry/) using port 5001. And it looks like kind changed from 5000 to 5001 precisely due to this reason:

* https://github.com/kubernetes-sigs/kind/pull/2621
* https://github.com/kubernetes/test-infra/pull/25134

Reply:
Updated the port to use 5001 on the host. But the port 5001 on the host maps to port 5000 on the container. This is because, the docker container uses the image `registry:2` which listens on port 5000.  This works for me. I am on version Ventura 13.4. Could you please confirm if the error is now resolved.

Reply:
> But the port 5001 on the host maps to port 5000 on the container.

@jubittajohn that's good. This is what we want. I just quickly chekced - registry is running fine now, but I'm still getting an error on an attempt to push the image:

```bash
  [FAILED] in [It] - /Users/mradchuk/workspace/src/github.com/operator-framework/operator-controller/test/operator-framework-e2e/operator_framework_test.go:318 @ 08/09/23 10:13:38.056
  << Timeline

  [FAILED] Unexpected error:
      <*errors.errorString | 0x1400014f4f0>:
      Error pushing bundle image localhost:5001/registry-operator-bundle:v0.1.0 with tag 0.1.0 : make[1]: Entering directory '/Users/mradchuk/workspace/src/github.com/operator-framework/operator-controller/test/operator-framework-e2e/registry-operator'
      make docker-push IMG=localhost:5001/registry-operator-bundle:v0.1.0
      make[2]: Entering directory '/Users/mradchuk/workspace/src/github.com/operator-framework/operator-controller/test/operator-framework-e2e/registry-operator'
      docker push localhost:5001/registry-operator-bundle:v0.1.0
      Getting image source signatures
      Copying blob sha256:dea1f3bba92bc840b14e6ae4e75f6d1538be9c7d11b13e27dadb691cd16c6a04
      Copying blob sha256:b6d433180825c5881b9565b9cbff353bace1926fe429c2cc4c67b75b203c8706
      Copying blob sha256:7e13d0a7484e3297efc0eb92610081b75a115f7a24b2ffde5e25bde7f6619634
      Error: trying to reuse blob sha256:b6d433180825c5881b9565b9cbff353bace1926fe429c2cc4c67b75b203c8706 at destination: pinging container registry localhost:5001: Get "https://localhost:5001/v2/": http: server gave HTTP response to HTTPS client
      make[2]: *** [Makefile:131: docker-push] Error 125
      make[2]: Leaving directory '/Users/mradchuk/workspace/src/github.com/operator-framework/operator-controller/test/operator-framework-e2e/registry-operator'
      make[1]: *** [Makefile:240: bundle-push] Error 2
      make[1]: Leaving directory '/Users/mradchuk/workspace/src/github.com/operator-framework/operator-controller/test/operator-framework-e2e/registry-operator'
      : exit status 2
      {
          s: "Error pushing bundle image localhost:5001/registry-operator-bundle:v0.1.0 with tag 0.1.0 : make[1]: Entering directory '/Users/mradchuk/workspace/src/github.com/operator-framework/operator-controller/test/operator-framework-e2e/registry-operator'\nmake docker-push IMG=localhost:5001/registry-operator-bundle:v0.1.0\nmake[2]: Entering directory '/Users/mradchuk/workspace/src/github.com/operator-framework/operator-controller/test/operator-framework-e2e/registry-operator'\ndocker push localhost:5001/registry-operator-bundle:v0.1.0\nGetting image source signatures\nCopying blob sha256:dea1f3bba92bc840b14e6ae4e75f6d1538be9c7d11b13e27dadb691cd16c6a04\nCopying blob sha256:b6d433180825c5881b9565b9cbff353bace1926fe429c2cc4c67b75b203c8706\nCopying blob sha256:7e13d0a7484e3297efc0eb92610081b75a115f7a24b2ffde5e25bde7f6619634\nError: trying to reuse blob sha256:b6d433180825c5881b9565b9cbff353bace1926fe429c2cc4c67b75b203c8706 at destination: pinging container registry localhost:5001: Get \"https://localhost:5001/v2/\": http: server gave HTTP response to HTTPS client\nmake[2]: *** [Makefile:131: docker-push] Error 125\nmake[2]: Leaving directory '/Users/mradchuk/workspace/src/github.com/operator-framework/operator-controller/test/operator-framework-e2e/registry-operator'\nmake[1]: *** [Makefile:240: bundle-push] Error 2\nmake[1]: Leaving directory '/Users/mradchuk/workspace/src/github.com/operator-framework/operator-controller/test/operator-framework-e2e/registry-operator'\n: exit status 2",
      }
  occurred
  In [It] at: /Users/mradchuk/workspace/src/github.com/operator-framework/operator-controller/test/operator-framework-e2e/operator_framework_test.go:318 @ 08/09/23 10:13:38.056
```

Reply:
@m1kola 
The issue is with using `podman`.

The error in pushing was due to `docker` being hard coded as the container runtime in the test. This is now changed to get the current container runtime using the environment variable `CONTAINER_RUNTIME`.  The default value of `CONTAINER_RUNTIME` defined in the `Makefile` is `docker`. Therefore the correct runtime has to be assigned to the variable `CONTAINER_RUNTIME` before calling the make target if `docker` is what is not being used. The test routine also assumes the runtime as `docker` if it is unable to retrieve the value of the environment variable.

But this is only a partial fix to the problem. With this change the test for `plain+v0` bundles will pass but for `registry+v1` will fail. This is because `registry+v1` uses `operator-sdk` support. Thus to mimic the user experience, the targets `bundle-build` and `bundle-push`  from the generated Makefile by `operator-sdk` tool, which has `docker` being hard coded as the container runtime, is used to build and push the bundle images. This could be marked as an issue and addressed when hard coding `docker` as container runtime is addressed by `operator-sdk`.

In conclusion, the current test will only work for `docker`.

Reply:
I have a symlink from `docker` to `podman` and Podman's CLI is compatible with docker. So `make operator-developer-e2e CONTAINER_RUNTIME=podman` and `make operator-developer-e2e` give me the same results.

It has to be something else.

I don't know how often (if at all) we will be running this test locally. It might be not worth spinning wheels on this (at least in context of this PR).